src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -240,23 +298,182 @@ inline DocumentStatisticsAnnotationData::DocumentStatisticsAnnotationData(Freque
+// |
+// |  Anonymous namespace
+// |
+// ----------------------------------------------------------------------
+namespace {
+
+inline Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap PruneTermFreqMap(Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap & termFrequency, 
+                                                                                    std::float_t minDf, 
+                                                                                    std::float_t maxDf, 
+                                                                                    std::float_t totalNumDocumentsFloat,
+                                                                                    nonstd::optional<std::uint32_t> topKTerms) {
+
+    using FrequencyMap = Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap;
+    
+    //trim by minDf and maxDf
+    if (minDf > 0.0f || maxDf < 1.0f) {
+        for (FrequencyMap::iterator termFrequencyIter = termFrequency.begin(); termFrequencyIter != termFrequency.end();) {
+            std::float_t const freq = termFrequencyIter->second / totalNumDocumentsFloat;
+
+            if (freq < minDf || freq > maxDf) {
+                termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            } else {
+                ++termFrequencyIter;
+            }       
+        }
+    }
+
+    //trim by topKTerms
+    if (!topKTerms.has_value() || termFrequency.empty())
+        return termFrequency;
+
+    std::uint32_t const & topKTermsValue(*topKTerms);
+
+    if (topKTermsValue > termFrequency.size())
+        throw std::invalid_argument("topKTermsUpperBound");
+
+    std::vector<std::uint32_t> frequencyVector;
+    frequencyVector.reserve(termFrequency.size());
+
+    std::transform(termFrequency.begin(), 
+                   termFrequency.end(), 
+                   std::back_inserter(frequencyVector), 
+                   [](auto const & kv) { 
+                       return kv.second; 
+                   });
+
+    //partial sort, ave O(n)
+    std::nth_element(frequencyVector.begin(), 
+                     frequencyVector.begin() + static_cast<int>(topKTermsValue) - 1, 
+                     frequencyVector.end(), 
+                     std::greater<std::uint32_t>());
+
+    std::uint32_t targetValue = frequencyVector[topKTermsValue - 1];
+    long long remainingTargetValue = std::count(frequencyVector.begin(), 
+                                                frequencyVector.begin() + static_cast<int>(topKTermsValue),
+                                                targetValue);
+
+    for (FrequencyMap::iterator termFrequencyIter = termFrequency.begin(); termFrequencyIter != termFrequency.end();) {
+        if (termFrequencyIter->second < targetValue) {
+            termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            continue;
+        } else if (termFrequencyIter->second == targetValue) {
+            if (remainingTargetValue-- <= 0) {
```

Comment:
This line is subtle, as the value becomes smaller regardless of the result of the condition (which happens to be OK since we only want this branch to execute when we have populated N items). Is there a way to rewrite this logic to more clearly express what is happening?